### PR TITLE
[GPU] scatter_update and shape_of shape infer + minor dyn flow fixes

### PR DESCRIPTION
**Pull request by vladimir-paramuzov**
### Details:
 - new shape infer impl for scatter_update and shape_of prims
 - minor fixes for gemm/gather and common primitive_inst

### CODE REVIEW

File: src/plugins/intel_gpu/include/intel_gpu/primitives/shape_of.hpp [hunk 1]
```diff
@@ -31,6 +31,17 @@ struct shape_of : public primitive_base<shape_of> {
...
+    /// @brief Constructs shape_of primitive.
+    /// @param id This primitive id.
+    /// @param input Input primitive id.
+    /// @param output_data_type type of output values. can be i32 and i64.
+    shape_of(const primitive_id& id,
+             const primitive_id& input,
+             const data_types output_data_type,
+             const padding& output_padding = padding())
+        : primitive_base(id, {input}, output_padding, optional_data_type{output_data_type})
+        , output_rank(0) {}
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/primitives/shape_of.hpp:43]: I think we can just merge those two APIs to one with default value.

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 2]
```diff
@@ -511,6 +511,8 @@ memory::ptr primitive_inst::allocate_output(engine& _engine, memory_pool& pool,
 
     std::function<bool(const program_node&)> user_requesting_mem_reuse_false = [&user_requesting_mem_reuse_false](const program_node& node) {
         for (auto& user : node.get_users()) {
+            if (user->is_dynamic())
+                continue;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:515]: I think even if one of the user is dynamic it should return true.

File: src/plugins/intel_gpu/src/graph/shape_of.cpp [hunk 3]
```diff
@@ -33,14 +33,30 @@ layout shape_of_inst::calc_output_layout(shape_of_node const& node, kernel_impl_
...
+template<typename ShapeType>
+std::vector<layout> shape_of_inst::calc_output_layouts(shape_of_node const& /*node*/, const kernel_impl_params& impl_param) {
+    auto prim = impl_param.typed_desc<shape_of>();
+
+    data_types output_dt = prim->output_data_type.value_or(data_types::i32);
+    if (impl_param.has_fused_primitives()) {
+        output_dt = impl_param.get_fused_output_layout().data_type;
+    }
+
+    auto in_shape = impl_param.get_input_layout(0).get<ShapeType>();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/shape_of.cpp:45]: should return dynamic layout if input rank is dynamic?

File: src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp [hunk 4]
```diff
@@ -37,14 +37,17 @@ void CreateElementwiseOp(Program& p, const std::shared_ptr<ngraph::Node>& op, cl
     auto inputPrimitives = p.GetInputPrimitiveIDs(op);
     std::string layerName = layer_type_name_ID(op);
 
-    auto outRank = op->get_output_shape(0).size();
+    auto out_pshape = op->get_output_partial_shape(0);
+    auto out_rank = out_pshape.size();
     for (size_t i = 0; i < inputPrimitives.size(); ++i) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp:42]: The entire loop of this can be applied only for static case.

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp:42]: This is exactly what is checked below: input_pshape.is_static() && out_pshape.is_static()

File: src/plugins/intel_gpu/src/graph/shape_of.cpp (hunk 3 above)
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/shape_of.cpp:45]: Dynamic rank won't be supported in the near future

File: src/plugins/intel_gpu/include/intel_gpu/primitives/shape_of.hpp (hunk 1 above)
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/shape_of.hpp:43]: We can merge for sure, but my idea is to separate api which can't be fixed immediately, and then just remove it once we migrate to new shape infer.

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 2 above)
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp:515]: Fixed

File: src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp (hunk 4 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp:42]: Yeah, so my question was, isnt it better to let the loop itself not to be entered if it is not static?

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp:42]: I don't see any big difference in your approach

Comment (yeonbok) [reply src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp:42]: Of course its not a critical change, but just for better readability and clearity to quickly check the static shape only part from the large point of view.

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp]: updated

**Review by yeonbok — APPROVED** [review]

**Pull request merged**